docs/design/persistent-storage.md

```diff
@@ -0,0 +1,235 @@
+
+```
+$ cluster/kubectl.sh get storage
+
+NAME                LABELS              MODES               RESOURCES
+_pv_vol_01          name=my-data        RWO,ROX             size=5gb,iops=30
+_pv_vol_02                              RWO,ROX             size=5gb,iops=30
+_pv_vol_03          name=other-data     RWO,ROX             size=25gb,iops=30
+
+```
+
+#### Request persistent storage for a pod
+
+Susie is happy to see storage that can be mounted many times (ReadOnlyMany).  That satisfies her load balancer requirement.  Susie creates an pod image with her number crunching code and requests persistence for her volume.
+
+Susie's YAML looks something like this for the pod that creates her data:
+
+
+```yaml
+
+id: creator-of-data
+kind: Pod
+namespace: ns
+apiVersion: v1beta1
+desiredState:
+  manifest:
+    version: v1beta1
+    id: creator-of-data
+    containers:
+      - name: numbercruncher
+        image: ns/makesomdata:v1
+        command:
+          - /run.sh
+        volumeMounts:
+          - name: data
+            mountPath: /some_data
+    volumes:
+      - name: data
+        persistenceRequirements:
+        	availableModes: readwriteonce | readonlymany
+        	currentMode: readwriteonce
+        	retentionPolicy: retainondelete
+        	resources:
+        		- size: 5
+        	selector:
+        		- name: my-persistent-data
+labels:
+  name: creator-of-data
+
+```
+
+#### Behind the curtain
+
+`PersistentStoreBinder` is watching for new pods with `PersistenceRequirements`.  The binder compares the request to all available stores and binds it to the closest match.
+
+To bind:
+
+* Set PersistentStore.Status.PodRef = pod that originally requested persistence
+* Set PersistentStore.Labels = PersistenceRequirements.Selector
+
+The intent is to bind a PersistentStore temporarily to a Namespace and give it labels others want to use when selecting it.  Query by label and filter by the namespace of the pod that originally requested persistence.
+
+The scheduler will wait to schedule a pod until the pod is bound with a persistent store.  Invert responsibility for host selection when scheduling pods with persistent stores.  Allow the controller to give advice to the scheduler, keeping volume rules contained in the provider. 
+
+Store controllers are watching bound pods to attach volumes for cloud providers.
+
+Kubelet waits to mount a filesystem until a cloud provider has finished attaching the volume.
+
+
+#### Surviving the pod lifecycle
+
+Susie's pod finishes its work and exits.  Because her retention policy was "RetainOnDelete", she is relying on Kubernetes to make the data available again to her in the same namespace using the same selector.
```

Comment:
This feels like something that should be handled when gracefully terminating a replication controller.